### PR TITLE
e2e: improve readiness probe checking test

### PR DESCRIPTION
**Pull request by fmuyassarov**
The name of the configMap file is missing .in at the end. Otherwise, instantiate function fails to find the file in the tree.
@klihub @jukkar PTAL.
Sorry I missed this earlier :)

### CODE REVIEW

**Review by klihub — CHANGES_REQUESTED** [review]
There is actually no reason to '$(instantiate ...) the configmap file in the test case if you do it like it is. There are no variable substitutions in the file. So you should be able to `vm-put-file` it directly to the test VM and run the test.

My bad, as I missed the added test case(s) in the latest incarnation of the PR and therefore failed to review it properly before it got merged. I apologize.

So anyway, I think if we keep the test like they are, it would be a less confusing/more proper fix to get rid of using instantiation without any substitutions.

However, the test case could use an instantiable ConfigMap, if you also tested the readiness transition in the other direction. IOW, break the ConfigMap first and wait for the pod to become not ready (or fail after a timeout)... which is what you already have. Then fix the ConfigMap and wait the the pod to declare readiness (again failing after a timeout)... which is missing from the tests.

If you update the test like that, then breaking and fixing the configuration can be done with instantiation and substituting something like `$GARBAGE` at the end of the file first with `foobar: xyzzy` and then with the empty string.

**Comment by jukkar** [comment]
> There is actually no reason to '$(instantiate ...) the configmap file in the test case if you do it like it is. There are no variable substitutions in the file. So you should be able to `vm-put-file` it directly to the test VM and run the test.

Yes, but if directly calling `vm-put-file` we would need to hard code the source path to the test script. This is why I asked Feruz to use the instantiate function so that we could avoid setting the path (because instantiate will do it for us). This works even if we do no variable substitution for the .in file.

**Comment by klihub** [comment]
> > There is actually no reason to '$(instantiate ...) the configmap file in the test case if you do it like it is. There are no variable substitutions in the file. So you should be able to `vm-put-file` it directly to the test VM and run the test.
> 
> Yes, but if directly calling `vm-put-file` we would need to hard code the source path to the test script. This is why I asked Feruz to use the instantiate function so that we could avoid setting the path (because instantiate will do it for us). This works even if we do no variable substitution for the .in file.

Okay. So let's do as we agreed during office day and what was also suggested above: let's add the missing half of the tests to check correct transitioning back from failed/no-ready state to ready one once the config map errors are fixed, and let's use an `ConfigMap` template with some `$EXTRA_FIELD` or similar placeholder using instantiation to generate first an incorrect then later a correct `ConfigMap`.

In addition to this, maybe we could consider later another PR which also teaches similar 'inheritance' to `vm-put-file` like other bits in the infra which look for files by walking up the directory from the test case dir to the test root dir until they find the file or fail in the end. That allows better sharing common files without having to use the $(instantiate $filename) workaround/trick that Feruz came up with.

**Comment by fmuyassarov** [comment]
@jukkar @klihub thanks for the suggestions above. I have now improved this PR and added two more commits with their description of what they are doing. PTAL.

**Review by jukkar — DISMISSED** [review]
LGTM

**Review by klihub — APPROVED** [review]
LGTM apart from a few nits below...

File: test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh
Comment (klihub) [inline test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: I think instead of the `sleep 5; kubectl get event...`-combo it would be better to do something like

`vm-command "kubectl -n kube-system wait --for=condition=Ready=true pods/$POD --timeout=60s"`

That would allow both a more relaxed timeout, yet not waste any time once the readiness transition has happened.

File: test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh
Comment (klihub) [inline test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh]: Ditto here... I think instead of the `sleep 10; kubectl get event...`-combo it would be better to do something like

`vm-command "kubectl -n kube-system wait --for=condition=Ready=false pods/$POD --timeout=60s"`

That would allow both a more relaxed timeout, yet not waste any time once the readiness transition has happened.

File: test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh
Comment (klihub) [inline test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh]: And ditto here... I think instead of the `sleep 5; kubectl get event...`-combo it would be better to do something like

`vm-command "kubectl -n kube-system wait --for=condition=Ready=true pods/$POD --timeout=60s"`

That would allow both a more relaxed timeout, yet not waste any time once the readiness transition has happened.

**Review by klihub — CHANGES_REQUESTED** [review]
@fmuyassarov A few nits about how to wait for and verify the expected Readiness state transitions. Otherwise LGTM.

File: test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh
Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: Actually I stopped using kubectl event in https://github.com/jukkar/nri-resmgr/pull/66/commits/3ca4397c11f3ca0d69a4c135dbae5e2e19eabc0a but I missed to update the same for balloons. 

But for the timeout, I think it is better to use the kubectl flag instead.

Comment (klihub) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: > Actually I stopped using kubectl event in [3ca4397](https://github.com/jukkar/nri-resmgr/commit/3ca4397c11f3ca0d69a4c135dbae5e2e19eabc0a) but I missed to update the same for balloons.
> 
> But for the timeout, I think it is better to use the kubectl flag instead.

But would that work as desired when used together with `kubect get {event|pod}` ? I think not. Those would return the current event or pod state without waiting for anything else than an response from the API server which will be pretty much immediate. That is very different from `kubectl wait` where you tell what condition you want kubectl to *wait for* and the timeout indicates the period after which `kubectl` should consider the condition/transition to have failed to happen.

If your idea was to optimize (IOW shorten) how fast the tests can succeed, then wouldn't it be better to simply increase the frequency of the readiness probe enough and use `kubectl wait` ?

Comment (jukkar) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: > vm-command "kubectl -n kube-system wait --for=condition=Ready=true pods/$POD --timeout=60s"

instead of this, we have `vm-wait-pod-regexp()` that does exacty that

Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: @klihub sorry you were right. I assumed that I could use --timeout with kubectl get pod, which was a wrong assumption. 
What you have proposed is much simpler and cleaner. I have updated the commit now accordingly. Thanks for the suggestion.

**Comment by fmuyassarov** [comment]
@klihub fixed issues. PTAL.

File: test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh
Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: > > vm-command "kubectl -n kube-system wait --for=condition=Ready=true pods/$POD --timeout=60s"
> 
> instead of this, we have `vm-wait-pod-regexp()` that does exacty that

I think that function has some limitations for our use case, for example `kubectl wait --timeout=${wait_t}s --for=condition=${wait}` line doesn't have true/false option, it only looks for condition but not the value. I can extend it and I just tried it but it is getting out of scope of this PR. TBH, I'm not super happy about using that function in here, because it is a wrapper which complicates a simple kubectl command. Perhaps it is useful in other scenarios but not sure here ?

Comment (klihub) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: @fmuyassarov But it would/should work if you use it like `wait=condition=Ready=false wait_t=5 vm-wait-pod-regexp $POD`...

Even the extra `kubectl get pods | grep ...` skulduggery which can't be skipped with any 'named option' should be fine (even though it is totally unnecessary in our case) and find the existing pod immediately.

That is then an entirely different matter if doing the wait indirectly using `vm-wait-pod-regexp` actually makes the test case easier or more difficult to understand. I'd claim that if someone is aware of `kubectl wait --for=condition=...` then it only makes things more difficult to read/understand.

IMO the only situation where `vm-wait-pod-regexp` probably helps in readability is, if you have a test case where you create a number of pods from one shell/subshell and want to wait for some condition of the yet to be created pods from another shell/subshell. Then you can do it with a single combined call, instead of first waiting for the pod to appear and then doing the wait.

Comment (jukkar) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: Althought I already +1, I think we should change this a bit as suggested by Krisztian,
to use the `vm-wait-pod-regexp()` function

Comment (klihub) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: I'm fine/LGTM with this either way...  `vm-wait-pod-regexp()` or vanilla `kubectl wait --for`. No strong opinions here. You guys can decide.

Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: I didn't notice that we can pass anything into the wait. Adding wait=Ready=true/false worked. Thanks

Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: sorry for delay @klihub @jukkar . I wasted my time by adding condition value incorrectly and searching for the issue in a wrong place. I have fixed it now. PTAL.

File: test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh
Comment (jukkar) [inline test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: Can't we use here the function from vm?

File: test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh
Comment (jukkar) [inline test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: ditto

File: test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh
Comment (fmuyassarov) [reply test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh]: @jukkar my bad, I again missed the balloons policy :) Fixed it now. Thanks for the catch.

**Review by klihub — APPROVED** [review]
LGTM

**Comment by fmuyassarov** [comment]
@jukkar PTAL it is ready from my side.

**Review by jukkar — APPROVED** [review]

**Pull request merged**
